# 260824 Numpy & Pandas 
[AI Literacy-prompt 실습(Lovable), Numpy&Pandas 특강] 

---

## 1. Today's Goal

- Numpy, Pandas 이해
- Vector/Matrix/Tensor 이해
- CLI 과제 제출 완료

In [ ]:
num1 = 10 + 3.5
num2 = int(3.99)
print(num1, num2)

---

## 2. Key Concepts

### Numpy
- 선형대수 기반의 Vector 계열, 배열로 작성
- 내부적으로 C로 구현되어 있음 -> 빠른 연산 가능
- 각 원소의 크기가 동일
     - index 0 -> 8 bit만큼 이동 -> index 1 -> 8 bit 이동 -> index 2 ....
- 단순한 배열 위주의 데이터 구조 -> 복잡한 데이터 분석이나 조작에 제한
- 딥러닝 주요 알고리즘에서 사용
- optimized for raw numerical computation and matrices
- 데이터 타입이 같음 --> 문자열 x, 숫자만 입력 가능(int/float)
- 1D, 2D, 3D ... N-dimensions
- 다차원 슬라이싱, reshape...
- 인덱스는 int
- dtype: 자료형 명시 ( ex> dtype=np.int64 또는 np.float64)

[배열]

1차원 배열 = `vector`

**2차원 배열 = `matrix`**

    - 행, 열 존재
    - [n번 행][n번 열]
    - [n번 행, n번 열]

3차원 배열 = `tensor`

`dtype=np.float64` = 자료형 명시



[리스트]
- 리스트 자체에는 객체를 가르키는 참조가 지정됨
- 실제 객체는 메모리에 다른 위치에 존재 가능
- 리스트 내부에는 참조들이 연속적으로 저장됨


[원소 크기]
- 비트(bit): 0.1
- 바이트(byte): 비트x8 ex) 1byte = 8bit.
- 부호비트(sign bit): 숫자가 양수/음수 판단하는 비트. 8비트에서 가장 왼쪽 비트 ex) 0 = 양수, 1 = 음수


In [ ]:
import numpy as np

# 1차원 배열
vec = np.array([10, 20, 30], dtype=np.int64)

# 2차원 배열
mat = np.array([
    [10, 20, 30],
    [40, 50, 60],
    [70, 80, 90]
])

print("1D Vector:", vec)
print("2D Matrix:\n", mat)

#### Numpy 슬라이싱/배열선택

- `matrix[행, 열]`
- `matrix[0:2, 1:3]` : 행(0~1), 열(1~2)
- `matrix[:,0]` : 행(0~끝), 열(0)  ==> 모든 행의 0번째 열 추출
- `matrix[1:, :2]` : 행(1~끝), 열(0~1)

#### Numpy 차원 변경, 축(Axis)
- `reshape(행, 열)`: 1차원 배열 -> 차원수 변경, -1=자동 결정
- `flatten()`: 2차원 -> 1차원 평탄화
- `arange(시작, 종료, 증감단위)`: 시작 이상~종료 미만 숫자 생성
- `axis`: 2차원 배열에서 연산 방향 설정
    - `axis=0`: 행 기준
    - `axis=1`: 열 기준

In [ ]:
vec = np.arange(1, 11)

mat = vec.reshape(2, -1) # vec 1D -> mat 2D

flat_mat_vected = mat.flatten() # mat 2D -> 1D

print(flat_mat_vected)

In [ ]:
print(mat)
sum_axis0 = mat.sum(axis=0)
print(sum_axis0)
sum_axis1 = mat.sum(axis=1)
print(sum_axis1)

#### Numpy Broadcasting
- 서로 크기가 다른 배열 간에 형상을 맞춰 자동 확장 연산 수행

**스칼라 브로드캐스팅**
- 모든 원소에 일괄 적용

**2D + 1D**
- 2d 각 행마다 vector 복제 및 연산 수행

In [5]:
import numpy as np
vec = np.array([1, 2, 3])
mat = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

print(vec)
# scalar broadcasting
mat_plus = mat + 5
print(mat_plus)


# 2d + 1d
mat_vec = mat + vec
print(mat_vec)

[1 2 3]
[[ 6  7  8]
 [ 9 10 11]
 [12 13 14]]
[[ 2  4  6]
 [ 5  7  9]
 [ 8 10 12]]


In [12]:
a = np.array([1, 2, 3])
b = np.array([[1], [2], [3]])
print(a)
print(b) 
print(a+b)
print(a*b)
print(a-b)
print(a/b)

[1 2 3]
[[1]
 [2]
 [3]]
[[2 3 4]
 [3 4 5]
 [4 5 6]]
[[1 2 3]
 [2 4 6]
 [3 6 9]]
[[ 0  1  2]
 [-1  0  1]
 [-2 -1  0]]
[[1.         2.         3.        ]
 [0.5        1.         1.5       ]
 [0.33333333 0.66666667 1.        ]]


#### Numpy 행렬 전치(T)와 행렬곱(@)

- `객체.transpose()`: {행, 열} ==> {열, 행}
    - `객체.T`


- `행렬곱 @`: 선형대수 규칙, **각 행과 각 열을 곱해서 더함** [가로 한 줄 × 세로 한 줄 → 곱한 다음 전부 더하기]






In [ ]:
mat = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

print(mat)
print("전치 행렬:\n", mat.transpose())
print("전치 행렬:\n", mat.T)

In [15]:
mat1 = np.array([[10, 20], [30, 40]])
mat2 = np.array([[1, 2], [3, 4]])

mat3 = mat1 @ mat2
print(mat1)
print(mat2)
print(mat3)

[[10 20]
 [30 40]]
[[1 2]
 [3 4]]
[[ 70 100]
 [150 220]]


$$
\begin{bmatrix}
10 & 20 \\
30 & 40
\end{bmatrix}
\begin{bmatrix}
1 & 2 \\
3 & 4
\end{bmatrix}
=
\begin{bmatrix}
10\times1 + 20\times3 & 10\times2 + 20\times4 \\
30\times1 + 40\times3 & 30\times2 + 40\times4
\end{bmatrix}
=
\begin{bmatrix}
70 & 100 \\
150 & 220
\end{bmatrix}
$$

### Pandas
- numpy 기반으로 만들어진 RDMS(관계형 데이터베이스) 
- 데이터를 관리하는 Table 구조
- 풍부한 데이터 조작 기능 제공 -> 복잡한 데이터 분석 작업에 용이
- DataFrame, Series 등의 데이터 구조
- CSV, Excel, SQL 등 다양한 파일 형식 쉽게 읽고 쓸 수 있음
- Numpy에 비해 상대적으로 더 많은 메모리 사용 -> 상대적으로 느림
- built for tabular data manipulation, spreadsheets, SQL-like queries
- 행 내에 다른 타입이 올 수 있음
- 2D
- rows / columns


#### pandas series, dataframe
- series = 1차원, index+data
- dataframe = 2차원, index+columns

In [ ]:
import pandas as pd
data = {
    "이름": ["김철수", "이영희", "박민수"],  
    "부서": ["개발", "기획", "개발"],
    "급여": [4500, 5200, 4800]
}

df = pd.DataFrame(data)
print(df) # 행 인덱스 0부터 시작 + column은 이름, 부서, 급여

#### Pandas 객체.shape | 객체.columns.tolist() | 객체.info() | 객체.describe()
- `객체.shape` : 형상(행, 열) 보여주기
- `객체.columns.tolist()`: 컬럼 목록 보여주기
- `객체.info()`: 요약 
    - dtype 섞일 경우 (ex. str+int ==> object)
- `객체.describe()`: 진단

In [ ]:
print("=" * 30)
print("행과 열:", df.shape) # 형상
print("=" * 30)
print("columns:", df.columns.tolist()) # column
print("=" * 30)
print("DataFrame 요약:")
print(df.info()) # 요약
print("=" * 30)
print("DataFrame 진단:\n", df.describe()) # 진단
print("=" * 30)

#### Pandas df.set_index | df.reset_index

- 열 추가 및 데이터 수정: `df['new column'] = df['old column'] * n`

- `df.set_index('column to be indexed')` : 인덱스로 만들 column 설정

- `df_to_restore_.reset_index()`: 인덱스로 만든 column 되돌리기 => 이전 dataFrame과는 다름


In [ ]:
import pandas as pd

# '보너스' column 추가
data = {
    "이름": ["김철수", "이영희", "박민수"],  
    "부서": ["개발", "기획", "개발"],
    "급여": [4500, 5200, 4800]
}

df['보너스'] = df['급여'] * 0.1

print(df)

In [ ]:
# column -> index
indexed_df = df.set_index('부서')
print(indexed_df)

In [ ]:
# index로 지정한 column dataframe reset
restored_df = indexed_df.reset_index()

print(restored_df)

#### Pandas Series & DataFrame 브로드캐스팅 연산

- `index = pd.Index([인덱스로 넣을 값], name='인덱스 이름 지정')`: DataFrame 지정시 인덱스 이름 + 데이터 지정 방법

In [ ]:
import pandas as pd

score_df = pd.DataFrame({
    '중간고사': [80, 95, 70],
    '기말고사': [85, 90, 75],
}, index=pd.Index(['김철수', '이영희', '박지민'], name='이름'))

print("=" * 30)
print("[시험 점수]")
print(score_df)
print("=" * 30)

col_mean = score_df.mean(axis=0)
print("[고사별 평균 점수]")
print(col_mean)
print("=" * 30)

row_mean = score_df.mean(axis=1)
print("[학생별 평균 점수]")
print(row_mean)
print("=" * 30)

deviation_df = score_df.sub(col_mean, axis=1)
print("[시험별 평균점수 - 학생 점수]")
print(deviation_df)
print("=" * 30)



***little bit of AI literacy - types of prompt***

- RTCCF
    - Role, Task, Context, Constraints, Form 


---

## 3. Practice 실습 코드

--- 
## 4. Daily Quest

> 1. 다음 중 1 이상 45 이하 범위에서 무작위 정수 하나를 반환하는 코드로 올바른 것은?

답: random.randint(1, 45)

> 2. 여러 값을 하나의 변수에 튜플 형태로 묶어 담는 것을 '패킹'이라고 합니다. 반대로 하나의 자료구조에 담긴 값들을 여러 개의 변수에 순서대로 나누어 담는 동작을 무엇이라고 하나요?

답: Packing

> 3. 리스트 numbers = [0, 1, 2, 3, 4, 5]가 있을 때, print(numbers[::2])의 실행 결과로 올바른 것은?

답: [0, 2, 4]

> 4. 부모 클래스로부터 물려받은 메서드를 자식 클래스의 요구사항에 맞게 새롭게 재정의하여, 호출 시 부모의 원본 메서드 대신 실행되도록 하는 객체지향 기법을 무엇이라 하나요?

답: class overriding

> 5. 다음 중 프로그램 실행 도중에 발생하는 '런타임 예외(Exception)'에 해당하는 것은?

답: C. 실행 중에 0으로 나누기 연산을 시도한 경우

5번 재정리

- SyntaxError: 문법 오류로 해석 불가
    - 오타
    - 들여쓰기 실수
    - 괄호 누락 등

- Exception(런타임 예외): 문법 완벽 but 실행중 연산 오류
    - 숫자 / 0
    - 리스트 범위 벗어난 인덱스 접근
    - 실행 전 식별 불가

--- 
## 5. Reflection

- CLI refactoring 잊지 않기
- Numpy와 Pandas가 헷갈려서 서로 비교하려 했는데, 검색 중 이 둘을 비교하는 건 마치 python과 C+를 비교하는 것과 같다는 글을 발견
- 데이터 정리 및 수정은 흥미롭지만 선형대수 개념이 나왔을 땐 당황스러웠다..
- AI prompt 기법이 존재한다는 건 흥미롭지만, 따로 배운다는 건 좀 지루하다고 
- python 놓지 않고 계속 연습하기! (while/for loop, if-if/if-elif 구분, CLI에 배웠던 library 적용해보기)